### ISSUE TITLE
Investigate the possibility of integration with JetBrains Gateway

### ISSUE THREAD

**Issue by azatsarynnyy** [eclipse/che#21412]
### Is your task related to a problem? Please describe

We observe some interest among the community in connecting to a DevWorkspace from the local IDEs. We need to provide a solution for the JetBrains IDEs.

### Describe the solution you'd like

Need to investigate how we can integrate Eclipse Che with JetBrains Gateway to make it possible to connect from a local JetBrains IDE to a remote DevWorkspace.

### Describe alternatives you've considered

It can be some PoC with just a base functionality.

### Additional context

`Getting started with remote development﻿` https://www.jetbrains.com/help/idea/remote-development-a.html

**Comment by vzhukovs** [comment]
There is a one restriction from using JetBrains Gateway - is running only paid versions of IDEs. It is not possible to run Community Editions using this technology.

**Comment by kstevena** [comment]
I don't know if it might be helpful, but I've been working on my side on a proof of concept, which looks to work, however this is quite dirty as I did not wanted to write a specific plugin for the IDE...

Here are some elements:

- JetBrains Gateway out of the box (no additional plugin installed) can connect to a remote environment through SSH, it uses that to make a port forwarding of the tcp port on which the server is listening.
To address this needs in addition of installing a pro version of the desired IDE from JetBrains, I've also installed an SSH server into the container. Which is started by the entrypoint script. The authorized keys of the user are automatically mounted through a configmap in the user namespace properly labelled & annotated.

- Endpoints can be of tcp type but won't be exposed outside of the cluster
As a workaround I made a controller watching for che/devworkspace generated services matching some specifications such as the name of a port port of the service being set to "ssh", then creating an associated service of type LoadBalancer.
In conjunction of that an instance of external-dns looking for those specific services, generating a templated dns record using the devworkspace Id

- The dashboard needs to redirect the user to a web page
As a workaround I made a small app to which the dashboard will redirect when the workspace will be ready and which will basically display and do a "window.open" of the JetBrains Gateway link. 
When the JetBrains Gateway has registered the "jetbrains-gateway://" type of link then it will open it smoothly.
JetBrains Gateway Link for ssh are of the following format:
`jetbrains-gateway://connect#idePath=%s&projectPath=%s&host=%s&port=%s&user=%s&type=ssh&deploy=false`
`IdePath`, `ProjectPath` and `user` are naturally known, but the host and ports are slightly different. However as in my PoC I used the hostname of the `CHE_DASHBOARD_URL` and the `DEVWORKSPACE_ID` to generate a DNS record and I relied on a static 22 port number it was then able to point to the correct host name for the SSH server.

As said this is far from being clean, but allowed me to demonstrate the feasability (and test if the experience is better than with projector or a vnc connection). Beyond the bad design, the current issues I have are the following:
- it requires one external IP per workspace (as I'm creating one service of type LoadBalancer per workspace requiring it), I assume it can be solved by configuring multiple ports on a single LoadBalancer entity ; not sure about another approach with a single ssh endpoint acting as a multiplexer or proxy such as [containerssh](https://github.com/containerssh/containerssh) or [sshpiper](https://github.com/tg123/sshpiper) or another one using an https proxy to be able to leverage TLS/SNI to point to the right ssh server within the cluster.
- when the workspace is fully started, the LoadBalancer might not yet be ready and so the SSH connectivity might not be yet working, so the user may have to wait a bit a do few retries with the JetBrains Gateway app.
